File: packages/python-sdk/prompttrace/tokenizer.py
import tiktoken
from typing import List, Dict, Any
from .types import TokenBreakdown

def estimate_tokens(messages: List[Dict[str, str]], model: str, output_text: str = '') -> TokenBreakdown:
    try:
        try:
            enc = tiktoken.encoding_for_model(model)
        except KeyError:
            enc = tiktoken.get_encoding('cl100k_base')

        system_tokens = 0
        user_tokens = 0
        history_tokens = 0

        for idx, msg in enumerate(messages):
            role = msg.get("role", "")
            content = msg.get("content", "")
            
            # Simple approximation (role + content + 4 format tokens)
            token_count = len(enc.encode(role + content)) + 4

            if role == 'system':
                system_tokens += token_count
            elif role == 'user' and idx == len(messages) - 1:
                user_tokens += token_count
            else:
                history_tokens += token_count

        output_tokens = len(enc.encode(output_text))

        return TokenBreakdown(
            systemTokens=system_tokens,
            userTokens=user_tokens,
            historyTokens=history_tokens,
            outputTokens=output_tokens
        )
    except Exception as e:
        print(f"[Prompttrace] Tokenizer error: {e}")
        return TokenBreakdown(0, 0, 0, 0)

File: packages/sdk/src/tokenizer.ts
import { get_encoding, encoding_for_model } from 'tiktoken';
import { TraceMessage, TokenBreakdown } from './types';

export function estimateTokens(messages: TraceMessage[], model: string, outputText: string = ''): TokenBreakdown {
  try {
    // Attempt to get encoding for the specific model, fallback to cl100k_base
    let enc;
    try {
      enc = encoding_for_model(model as any);
    } catch {
      enc = get_encoding('cl100k_base');
    }

    let systemTokens = 0;
    let userTokens = 0;
    let historyTokens = 0;

    messages.forEach((msg, index) => {
      // Simple approximation for tokens (role + content)
      const tokenCount = enc.encode(msg.role + msg.content).length + 4; // Add format tokens

      if (msg.role === 'system') {
        systemTokens += tokenCount;
      } else if (msg.role === 'user' && index === messages.length - 1) {
        userTokens += tokenCount;
      } else {
        historyTokens += tokenCount;
      }
    });

    const outputTokens = enc.encode(outputText).length;
    
    // Remember to free the encoding!
    enc.free();

    return {
      systemTokens,
      userTokens,
      historyTokens,
      outputTokens
    };
  } catch (error) {
    console.error('[Prompttrace] Tokenizer error:', error);
    return {
      systemTokens: 0,
      userTokens: 0,
      historyTokens: 0,
      outputTokens: 0
    };
  }
}
